EIPS/eip-6963.md

```diff
@@ -46,6 +46,7 @@ Provider Discovery Library: A library or piece of software that assists a DApp t
 Each wallet provider will be announced with the following interface `EIP6963ProviderInfo`. The values in the `EIP6963ProviderInfo` MUST be used as follows:
 
 - **`uuid`** - a globally unique of the wallet provider that MUST be (UUIDv4[^rfc4122] compliant) to uniquely distinguish different [EIP-1193](./eip-1193.md) provider sessions. The cryptographic uniqueness provided by UUIDv4[^rfc4122] guarantees that two independent `EIP6963ProviderInfo` objects can be separately identified.
+- **`rdns`** - an optional identifier that is used to optimistically identify the wallet by the Decentralized Application. If supplied by the wallet it MUST utilize an reverse DNS notation identifier that matches the domain name of the wallet provider.
```

Comment:
Move this below "name" and "icon" since they are required and this one is optional

Reply:
Wonder if this should be worded as RECOMMENDED (but optional) to make my life easier so us library authors don’t have to follow up with wallets because their wallet doesn’t work with RainbowKit, ConnectKit, etc…

Reply:
Won’t die on the hill though, but it’s just a caution. I’m grateful for this being here in the first place 😂

Reply:
Review closely, I added a lot of detail here resolving as many requests/comments as possible!
```suggestion
- **`rdns`** - a RECOMMENDED optional string that is used by the Decentralized Application to optimistically identify the kind of wallet being injected. If supplied by the wallet, it MUST contain a valid lowercase domain name in reverse order (e.g. `tld.domain.subdomain`) that is under the control of the developer of that wallet. It SHOULD NOT be supplied by any other party.
```

Reply:
As long as it's defined as OPTIONAL then I don't see an issue to also mention that it's RECOMMENDED

Reply:
RECOMMENDED is equivalent to SHOULD in RFC 2119. OPTIONAL is equivalent to MAY. My opinion is that we should stick with this being a MAY. For more rationale why I'm of this opinion see the final two paragraphs of this https://github.com/ethereum/EIPs/pull/7167#issuecomment-1619313354

Reply:
> Review closely, I added a lot of detail here resolving as many requests/comments as possibleI

I'm a +1 to this overall with the exception of using RECOMMENDED. Can you edit it to say `a string that MAY be used...`

Reply:
I agree with "MAY" implying that `rdns` may be used by the DApp, but I definitely think that we should include a "SHOULD" for Wallet Providers implementing `rdns`: "a RECOMMENDED string that MAY be used by the Decentralized Application".

If one of the top Wallets do not implement `rdns`, this will seriously jeopardize the connection UX of the DApp if they are using a wallet connection library. 

To put as an analogy, WalletConnect v2 has the concept of "optional chains", which only just a month before v1 cutoff, only one wallet had implemented the "optional chains" implementation. Optional chains are important, as libraries need it to create a working multichain experience. As a result, WalletConnect is advocating for libraries to lean more on the "optional chains" implementation. Can get into more detail about that, but just want to get to the point: if we are labelling this property with an "OPTIONAL" term, then I can see myself right now scrambling to get all these wallets to implement `rdns` (just like optional chains in WCv2) so that they can work with wagmi, RainbowKit, ConnectKit, Web3Modal, etc...

I am okay with this being an optional property (still feel strongly about required but if the consensus thinks otherwise, i concur), but at least, we need to label it as "a string that SHOULD be provided..."

Reply:
I think Kyle is right there that we SHOULD use all these capitalized verbs in the RFC2119 sense, not in the contemporary colloquial English usage of the words :D 

Here's the RFC2119 version of the edit:

```suggestion
- **`rdns`** - an OPTIONAL string that is used by the Decentralized Application to optimistically identify the kind of wallet being injected. It MAY be supplied by the wallet to facilitate smoother user experience but SHOULD NOT be supplied by any other party. If supplied by the wallet, it MUST contain only a valid lowercase domain name in reverse order (e.g. `com.walletbrand.extension`) that is under the control of the developer of that wallet.
```

Reply:
Or, if we wanted to signal to the wallet developer that this OPTIONAL isn't really realistically optional, we can sneak in a little rhetorical oomph in modifying clauses:

```suggestion
- **`rdns`** - an OPTIONAL string that is used by the Decentralized Application to optimistically identify the kind of wallet being injected to enable better end-user experience. It MAY be supplied by the wallet to opt in to this mechanism but SHOULD NOT be supplied by any other party. If supplied by the wallet, it MUST contain only a valid lowercase domain name in reverse order (e.g. `com.walletbrand.extension`) that is under the control of the developer of that wallet.
```

Reply:
> If one of the top Wallets do not implement rdns, this will seriously jeopardize the connection UX of the DApp if they are using a wallet connection library.

Can you specify how we'll jeopardize the connection UX with greater detail here so I can understand your concerns better? From what I can tell today, eip6963.org is working fine for discovery when I render non-existent wallets without the providerId property. Also, how is rendering useful if the provider object isn't available yet from the wallet? For example, in the video below how would it be useful for the Enkrypt wallet to appear if it's not actually available?

As you can see in this video, even with Enkrypt providing a `providerId` and the other 2 not, we're unable to disambiguate between the two as the spec currently because the implementation isn't error handling for the backwards compatibility aspect. This goes to show this is more about implementations handling different cases properly than it is about interfaces including properties. The issue here is no longer discovery, it's purely about implementation branching logic on the DApp side from what I can tell.

https://github.com/ethereum/EIPs/assets/23125059/2c77007e-605a-48b7-aff2-9d004c764184

Reply:
For all the reasons in my original PR. On top of that, without a deterministic identifier, it will not be trivially possible to sync up injected wallets with their corresponding WalletConnect wallet, which would cause duplication there, and is probably one of the more important cases I should have listed in the original PR.

> Also, how is rendering useful if the provider object isn't available yet from the wallet?

Not all wallets inject themselves immediately. Some could inject themselves a few seconds after. A DApp could provide a better experience here by displaying a loading indicator rather than nothing, and confusing the end-user. I thought we discussed this property being an extension point so that DApp libraries can do this? There is an error scenario here regarding wallet removal (an edge case), and that can be addressed accordingly.

Reply:
> For all the reasons in my original PR.

I believe you're referring to these [here](https://github.com/ethereum/EIPs/pull/7167/files#diff-927530d6d340560a972fe4fc1b91d7833e7a959089cc9c9d2baee54dd128c410R239-R241) is that correct. Can you show me a demo of these issues so we can properly assess the tradeoffs here between including and not including this? I'm hoping we can move past using hypotheticals and let the code determine the direction here. So far, what I'm seeing is that the UX is determined based on implementation choices on the DApp side rather than the inclusion of the `providerId` as I showed in the video above. 

Right now we're talking past each other because I'm coming at this from the perspective of knowing how text can be misinterpreted 4 years down the road and knowing that the spec text can't be updated after this occurs. I think (correct me if I'm wrong) you're coming at this from the perspective that if a bug occurs then people file an issue and all can be fixed. Implementations can be updated easily but specs cannot. Once a spec is finalized it can't be undone or modified. This is the difference between producing an open standard and producing an open source implementation. When you get a spec wrong it means all the legacy stuff has to stay in place which is the reason `eth_signTypedData` has 4 different versions. When an implementation is wrong, it get's fixed and versions get bumped and all is fine. That's why I'm trying to force us to identify and handle all of the bugs now before we finalize the spec.

> On top of that, without a deterministic identifier, it will not be trivially possible to sync up injected wallets with their corresponding WalletConnect wallet

I think that's a feature not a bug since are still two separate Wallets with two independent sets of state (e.g. loading a private key in my Desktop browser does not automatically load it into my mobile wallet). My mobile browser is not the same as my desktop browser either and I need to configure each of them independently. A worse scenario in my eyes is if two independent wallets get mismatched because of one of the errors I highlighted in this [comment](https://github.com/ethereum/EIPs/pull/7256#discussion_r1252359386).

> Not all wallets inject themselves immediately. Some could inject themselves a few seconds after. A DApp could provide a better experience here by displaying a loading indicator (before the provider hydrates) rather than nothing, and confusing the end-user.

Would it not confuse the end user to have a wallet appear that's been deleted as well? For example, today if that happens we automatically render the page to download the wallet, but the code path you're suggesting would assume the wallet is already available so is unlikely to render that page unless we define it to (which I don't think should be the case - that's an implementation detail in my eyes).

Reply:
Resolving this, we've opted to go with a REQUIRED usage of this property